src/v/storage/segment_utils.cc

```diff
@@ -595,6 +605,33 @@ ss::future<> transfer_segment(
 
     // clean up replacement segment
     co_await from->remove_persistent_state();
+
+    co_return std::move(locks);
+}
+
+ss::future<std::vector<ss::rwlock::holder>> write_lock_segments(
+  std::vector<ss::lw_shared_ptr<segment>>& segments,
+  ss::semaphore::clock::duration timeout,
+  int retries) {
+    vassert(retries >= 0, "Invalid retries value");
```

Comment:
why not `ss::parallel_for_each` ? i don't know how to express that w/ coros.

Reply:
Could that deadlock (until timer exceeds)? Shouldn't locks always be taken in the same order?

Reply:
yup, until timeout for sure.

Reply:
very rare for adjacent segments tho. remember that these are usually the begining of the log.

Reply:
https://github.com/vectorizedio/redpanda/compare/11070fee3ef2d6b5a6bd317e12f9de9101abf978..0af49b07b1c8d61ea3d2f2d46dcc2192f9cce1b5

@senior7515 yes, forgot to parallelize that. fixed that.

@BenPope totally. we use a timeout here to prevent that from persisting too long, but this is no silver bullet either from increasing latency. a totally different approach to locking and protecting state is needed in the storage system but that'll have to be for another day.